crates/control/src/search_suggestor.rs

```diff
@@ -0,0 +1,175 @@
+use color_eyre::Result;
+use context_attribute::context;
+use framework::{AdditionalOutput, MainOutput};
+use nalgebra::{DMatrix, Isometry2, Point2};
+use serde::{Deserialize, Serialize};
+use types::{
+    ball_position::{BallPosition, HypotheticalBallPosition},
+    field_dimensions::FieldDimensions,
+    parameters::SearchSuggestorParameters,
+};
+
+#[derive(Deserialize, Serialize)]
+pub struct SearchSuggestor {
+    heatmap_dimensions: (usize, usize),
+    heatmap: DMatrix<f32>,
+}
+
+#[context]
+pub struct CreationContext {
+    field_dimensions: Parameter<FieldDimensions, "field_dimensions">,
+    search_suggestor_configuration: Parameter<SearchSuggestorParameters, "search_suggestor">,
+}
+
+#[context]
+pub struct CycleContext {
+    search_suggestor_configuration: Parameter<SearchSuggestorParameters, "search_suggestor">,
+    ball_position: Input<Option<BallPosition>, "ball_position?">,
+    invalid_ball_positions: Input<Vec<HypotheticalBallPosition>, "invalid_ball_positions">,
+    robot_to_field: Input<Option<Isometry2<f32>>, "robot_to_field?">,
+    field_dimensions: Parameter<FieldDimensions, "field_dimensions">,
+    heatmap: AdditionalOutput<DMatrix<f32>, "ball_search_heatmap">,
+}
+
+#[context]
+#[derive(Default)]
+pub struct MainOutputs {
+    pub suggested_search_position: MainOutput<Option<Point2<f32>>>,
+}
+
+impl SearchSuggestor {
+    pub fn new(_context: CreationContext) -> Result<Self> {
+        let heatmap_dimensions = (
+            _context.field_dimensions.length.round() as usize
+                * _context.search_suggestor_configuration.cells_per_meter,
+            _context.field_dimensions.width.round() as usize
+                * _context.search_suggestor_configuration.cells_per_meter,
+        );
+        Ok(Self {
+            heatmap_dimensions,
+            heatmap: DMatrix::from_element(heatmap_dimensions.0, heatmap_dimensions.1, 0.0),
+        })
+    }
+
+    pub fn cycle(&mut self, mut context: CycleContext) -> Result<MainOutputs> {
+        self.update_heatmap(
+            context.ball_position,
+            context.invalid_ball_positions,
+            context.robot_to_field.copied(),
+            context.search_suggestor_configuration.cells_per_meter,
+            context.search_suggestor_configuration.heatmap_decay_factor,
+        );
+        let maximum_heat_heatmap_position = self.heatmap.iamax_full();
+        let mut suggested_search_position: Option<Point2<f32>> = None;
+        if self.heatmap.get(maximum_heat_heatmap_position).is_some() {
```

Comment:
How can the index of the maximum index be out of bounds? In my understanding this cannot happen. If this is correct, use normal array indexing syntax `[...]`

Reply:
You are right. I changed it to normal array indexing syntax